### DISCUSSION TITLE
song mode or channel mode features

### DISCUSSION THREAD

**Discussion by BilldarBagdar** [markusschloesser/MackieC4_P3#62] [Ideas]
starting to actually mess around "music making" instead of "programming" again.  So far, the script is working so well, I haven't noticed any issues with "existing features".  I'm only noticing "missing features" I'd like to implement.

I miss a way to toggle between "clip and device view" (Shift + Tab) and "session and arrangement view" (Tab) on this controller.  My APC40 models have dedicated buttons for the "clip and device view" toggle, and of course I can use the keyboard shortcuts.  But depending on the way my "studio chair" happens to be oriented, the C4 can be the most convenient "reach" (if you know what I mean).

We can either (or both) map a couple of encoder button clicks in "song mode" to toggle between the two views, or map a couple of other "click events" in any mode, something like maybe The "Option" or "Lock" button would toggle "clip and device view" and the "Alt" or "Spot Erase" button would toggle "session and arrangement view" across all modes, or whatever. (like the "Split" button toggles 3 LEDS, so depending on which "Split LED" is illuminated, some other button toggles between "clip and device view" or "session and arrangement view")

**Comment by markusschloesser** [comment]
yep, I've been enjoying working with the C4 immensely for the last couple of days and haven't even used it for the purpose for which I initially bought it (remote controlling the Yamaha FS1R). It's fucking awesome 😊😊
In fact my search for a VESA mount/arm has intensified so that I can use the C4 directly in front of the (3) screens.
That's also why a clip/device switch is not THAT important for me, but I agree, a mode which features stuff like this will be helpful.

As I am currently trying to implement an undo/redo functionality, let's make a list here of what we want to have on it:

- [x] clip / Device view toggle (Shift + Tab)
- [x] session / arrangement toggle (Tab)
- [x] BTA / Back to arrangement
- [x] Follow toggle
- [x] open / close browser
- [x] loop on/off
- [x] unsolo all
- [x] unmute all
- [ ] Undo / Redo
- [ ] metronome on/off/only while recording
- [ ] capture midi
- [ ] record quantization on/off
- [ ] save
- [ ] scroll through track in arrangement via vpot
- [ ] change BPM

Feel free to append!

**Reply by markusschloesser** [reply]
I'm gonna copy paste code here per feature

**Reply by markusschloesser** [reply]
probably makes sense to use transport.py from the v2 frameworks and a couple of others

**Comment by markusschloesser** [comment]
def __toggle_follow(self):
        self.song().follow_song = not self.song().follow_song

**Comment by markusschloesser** [comment]
def __toggle_loop(self):
        self.song().loop = not self.song().loop

**Comment by markusschloesser** [comment]
toggle arrange/session mode
```
    def __toggle_session_arranger_is_visible(self):
        if self.application().view.is_view_visible('Session'):
            if self.shift_is_pressed():
                self.application().view.focus_view('Session')
            else:
                self.application().view.hide_view('Session')
        elif self.shift_is_pressed():
            self.application().view.focus_view('Arranger')
        else:
            self.application().view.hide_view('Arranger')
```

**Comment by markusschloesser** [comment]
toggle clip / Device view
```
    def __toggle_detail_sub_view(self):
        if self.application().view.is_view_visible('Detail/Clip'):
            if self.shift_is_pressed():
                self.application().view.focus_view('Detail/Clip')
            else:
                self.application().view.show_view('Detail/DeviceChain')
        elif self.shift_is_pressed():
            self.application().view.focus_view('Detail/DeviceChain')
        else:
            self.application().view.show_view('Detail/Clip')
```

**Comment by markusschloesser** [comment]
show / hide browser
```
    def __toggle_browser_is_visible(self):
        if self.application().view.is_view_visible('Browser'):
            if self.shift_is_pressed():
                self.application().view.focus_view('Browser')
            else:
                self.application().view.hide_view('Browser')
        else:
            self.application().view.show_view('Browser')
```

**Comment by markusschloesser** [comment]
BTA
```
    def __toggle_back_to_arranger(self):
        self.song().back_to_arranger = not self.song().back_to_arranger
```

**Comment by markusschloesser** [comment]
(h)unsolo all 🤓
```
@unsolo_all_button.pressed
def unsolo_all_button(self, button):
    for track in tuple(self.song().tracks) + tuple(self.song().return_tracks):
        if track.solo:
            track.solo = False
```

**Comment by markusschloesser** [comment]
nearly done with the first 8 vpots, works but needs displaying

**Comment by markusschloesser** [comment]
@BilldarBagdar see latest commit in shared-dev 😁😉
Encoder LED ring doesn't work perfectly yet (no listeners), but it's a start, and something I could do
Edit: I'm talking about "Function" mode btw

**Comment by BilldarBagdar** [comment]
sweet.  I will check it out. 

I got a nice little mounting arm (I think it is VESA configured) from Sweetwater last fall.  It's a Gator brand, and they advertise it can support 3 - 8 kg, but the C4 has some mass to it, is it over 8 kg?    I also recently bought a Gator brand "utility stand" that has VESA mounting configured for the way the "table" attaches to the stand.  I can easily see how someone could mount monitors or whatever on stands like that as part of the show.

**Comment by BilldarBagdar** [comment]
Nice work.  I opened Live before I had turned on midi devices.  So I shut down Live and powered on the devices, and the C4 LCDs seemed to "power on" with the displays showing "Ableton Live" "device is offline".  Usually the top LCD shows "Mackie Control C4 Pro" for a while.  I saw that message in the code you committed, so I was mostly surprised that the C4 made a midi remote connection with Live after powering on fast enough for Live to "destroy it" while shutting down.

I don't know if you are ready to commit to having a release version, but I would say it's definitely time for a master branch update.  I don't necessarily mean any kind of immediately, I just mean the current shared-dev script version pretty much kicks ass over the master branch version right now.  There isn't really any reason to wait.  What if someone finds the repo and downloads master before this shared-dev (or any future) merges into master?  Even if you are still working on something half finished, we can merge that into master when it is ready.  (And whatever bugs are still in shared-dev were already in master anyway)

I'm going to look into the encoder rings and how they should "really" work.  That "ring feedback" works correctly when there is a DeviceParameter object involved,  The sends and volume do "wrap", and the Pan does "Spread", and "parameter encoders" do "wrap".  I think they all often start out as "single dot".  But I can't seem to get the encoder rings to just fully light up or fully shut off in response to button presses.

**Reply by markusschloesser** [reply]
> Nice work. 

Thanks, cleaning up stuff, even if it's minor visual things feels good. (and closing issues)

> I opened Live before I had turned on midi devices. So I shut down Live and powered on the devices, and the C4 LCDs seemed to "power on" with the displays showing "Ableton Live" "device is offline". Usually the top LCD shows "Mackie Control C4 Pro" for a while. I saw that message in the code you committed, so I was mostly surprised that the C4 made a midi remote connection with Live after powering on fast enough for Live to "destroy it" while shutting down.

yep, I noticed that, when I implemented the "turn vpot ring led off" thing, the destroy function does not only get called when closing Live, but also when loading a track etc.

> 
> I don't know if you are ready to commit to having a release version, but I would say it's definitely time for a master branch update. I don't necessarily mean any kind of immediately, I just mean the current shared-dev script version pretty much kicks ass over the master branch version right now. There isn't really any reason to wait. What if someone finds the repo and downloads master before this shared-dev (or any future) merges into master? Even if you are still working on something half finished, we can merge that into master when it is ready. (And whatever bugs are still in shared-dev were already in master anyway)

agreed! Only wann fix some minor things tonight in Function mode.
Do we "merge" or is there also a "replace" code? Sooo much stuff has changed.

> 
> I'm going to look into the encoder rings and how they should "really" work. That "ring feedback" works correctly when there is a DeviceParameter object involved, The sends and volume do "wrap", and the Pan does "Spread", and "parameter encoders" do "wrap". I think they all often start out as "single dot". But I can't seem to get the encoder rings to just fully light up or fully shut off in response to button presses.

I wondered about that as well, but that's how it was already implemented for other things. So I thought that you had liked it that way. I personally would most likely prefer full unlight / full lights on.

**Reply by BilldarBagdar** [reply]
Yes - I like the idea of "blinking" from full off to full on with toggles.  The LED ring is an area I "always knew" was crappy and didn't really work, but never got to the bottom of what the heck is going on.

I suspect we need a way to "map midi for feedback" without also mapping a "device parameter"

This works correctly for feedback to the LED rings: 
```
            feedback_rule = Live.MidiMap.CCFeedbackRule()  
            feedback_rule.channel = 0  
            feedback_rule.cc_no = self.__vpot_cc_nbr  
            display_mode_cc_base = encoder_ring_led_mode_cc_values[self.__v_pot_display_mode][0]
            range_end = encoder_ring_led_mode_cc_values[self.__v_pot_display_mode][1] - display_mode_cc_base
            feedback_rule.cc_value_map = tuple([display_mode_cc_base + x for x in range(range_end)])
```

But you can't seem to map for "midi feedback" (to the LED rings) this way.  You seem to need to do the "full enlightened poti" way.  But that doesn't seem to work either...

**Comment by markusschloesser** [comment]
I opened an issue for Function mode ( #65) , feel free to append!
I will also "hide" stuff here, that's now in the issue
